README.md
vuettext
==============

Gettext tools to improve your internationalization workflow with Vue-i18n

[![Version](https://img.shields.io/npm/v/vuettext.svg)](https://npmjs.org/package/vuettext)
[![CircleCI](https://img.shields.io/circleci/project/github/shimarulin/vuettext.svg)](https://circleci.com/gh/shimarulin/vuettext)
[![Downloads/week](https://img.shields.io/npm/dw/vuettext.svg)](https://npmjs.org/package/vuettext)
[![oclif](https://img.shields.io/badge/cli-oclif-brightgreen.svg)](https://oclif.io)
[![License](https://img.shields.io/npm/l/vuettext.svg)](https://github.com/shimarulin/vuettext/blob/master/package.json)

<!-- toc -->
* [Usage](#usage)
* [Commands](#commands)
<!-- tocstop -->
# Usage
<!-- usage -->
```sh-session
$ npm install -g vuettext
$ vuettext COMMAND
running command...
$ vuettext (-v|--version|version)
vuettext/0.6.0 linux-x64 node-v11.6.0
$ vuettext --help [COMMAND]
USAGE
  $ vuettext COMMAND
...
```
<!-- usagestop -->
# Commands
<!-- commands -->
* [`vuettext help [COMMAND]`](#vuettext-help-command)
* [`vuettext msgfmt`](#vuettext-msgfmt)
* [`vuettext msginit FILE`](#vuettext-msginit-file)
* [`vuettext msgmerge`](#vuettext-msgmerge)
* [`vuettext xgettext PATTERN`](#vuettext-xgettext-pattern)

## `vuettext help [COMMAND]`

display help for vuettext

```
USAGE
  $ vuettext help [COMMAND]

ARGUMENTS
  COMMAND  command to show help for

OPTIONS
  --all  see all commands in CLI
```

_See code: [@oclif/plugin-help](https://github.com/oclif/plugin-help/blob/v2.1.4/src/commands/help.ts)_

## `vuettext msgfmt`

Convert *.po messages to vue-i18n format

```
USAGE
  $ vuettext msgfmt

OPTIONS
  -h, --help           show CLI help

  -i, --input=input    (required) [default: i18n/messages/*.po] Glob pattern to specify *.po files. Expected string or
                       separated comma strings
                       Needs to be surrounded with quotes to prevent shell globbing.
                       Guide to globs: https://github.com/isaacs/node-glob#glob-primer

  -o, --output=output  (required) [default: i18n/locales] Path to the locales directory

  --ignore=ignore      Glob pattern to specify ignored files.
                       Expected string or separated comma strings
```

_See code: [src/commands/msgfmt.ts](https://github.com/shimarulin/vuettext/blob/v0.6.0/src/commands/msgfmt.ts)_

## `vuettext msginit FILE`

Initialize language translation (*.po file for selected language)

```
USAGE
  $ vuettext msginit FILE

ARGUMENTS
  FILE  [default: i18n/messages.pot] Input *.pot file

OPTIONS
  -h, --help
      show CLI help

  -l, --locale=locale
      language code of the language. This can be in one of three forms:
      - 'll', an ISO 639 two-letter language code (lowercase). See Language Codes for the list of codes.
      - 'll_CC', where 'll' is an ISO 639 two-letter language code (lowercase) and 'CC' is an ISO 3166
         two-letter country code (uppercase). The country code specification is not redundant: Some languages
         have dialects in different countries. For example, 'de_AT' is used for Austria, and 'pt_BR' for Brazil.
         The country code serves to distinguish the dialects. See Language Codes and Country Codes for
         the lists of codes.
      - 'll_CC@variant', where 'll' is an ISO 639 two-letter language code (lowercase), 'CC' is an ISO 3166
         two-letter country code (uppercase), and 'variant' is a variant designator. The variant designator
         (lowercase) can be a script designator, such as 'latin' or 'cyrillic'.

  -o, --output=output
      Path to output file
```

_See code: [src/commands/msginit.ts](https://github.com/shimarulin/vuettext/blob/v0.6.0/src/commands/msginit.ts)_

## `vuettext msgmerge`

Merge outdated messages (*.po files) and updated *.pot template

```
USAGE
  $ vuettext msgmerge

OPTIONS
  -d, --def=def    (required) [default: i18n/locales/*.po] Glob pattern to specify message *.po files. Expected string
                   or separated comma strings
                   Needs to be surrounded with quotes to prevent shell globbing.
                   Guide to globs: https://github.com/isaacs/node-glob#glob-primer

  -h, --help       show CLI help

  -r, --ref=ref    (required) [default: i18n/messages.pot] Path to updated reference file (*.pot template)

  --ignore=ignore  Glob pattern to specify ignored files
                   Expected string or separated comma strings
```

_See code: [src/commands/msgmerge.ts](https://github.com/shimarulin/vuettext/blob/v0.6.0/src/commands/msgmerge.ts)_

## `vuettext xgettext PATTERN`

Extract Vue-i18n strings from *.js and *.vue files to *.pot file

```
USAGE
  $ vuettext xgettext PATTERN

ARGUMENTS
  PATTERN  [default: **/*.{js,vue}] Glob pattern to specify files to be extracting
           Vue-i18n strings. Expected string or separated comma strings
           Needs to be surrounded with quotes to prevent shell globbing.
           Guide to globs: https://github.com/isaacs/node-glob#glob-primer

OPTIONS
  -h, --help           show CLI help
  -o, --output=output  (required) [default: i18n/messages.pot] Path to output file

  --ignore=ignore      Glob pattern to specify ignored files
                       Expected string or separated comma strings
```

_See code: [src/commands/xgettext.ts](https://github.com/shimarulin/vuettext/blob/v0.6.0/src/commands/xgettext.ts)_
<!-- commandsstop -->


CHANGELOG.md
# [0.6.0](https://github.com/shimarulin/vuettext/compare/v0.5.1...v0.6.0) (2019-04-05)


### Bug Fixes

* **circleci:** broken CircleCI config ([fd8a300](https://github.com/shimarulin/vuettext/commit/fd8a300))


### Features

* Add class fields support ([#19](https://github.com/shimarulin/vuettext/issues/19)) ([44e8f60](https://github.com/shimarulin/vuettext/commit/44e8f60))
* add support to dynamic imports ([a08f146](https://github.com/shimarulin/vuettext/commit/a08f146)), closes [#14](https://github.com/shimarulin/vuettext/issues/14)
* Add support to parse variable declarations ([322449c](https://github.com/shimarulin/vuettext/commit/322449c)), closes [#16](https://github.com/shimarulin/vuettext/issues/16)



## [0.5.1](https://github.com/shimarulin/vuettext/compare/v0.5.0...v0.5.1) (2019-01-27)



# [0.5.0](https://github.com/shimarulin/vuettext/compare/v0.4.0...v0.5.0) (2019-01-27)


### Features

* **msgfmt:** add msgfmt command ([59a9e32](https://github.com/shimarulin/vuettext/commit/59a9e32))
* **msgfmt:** add nested messages to msgfmt output ([3ee359f](https://github.com/shimarulin/vuettext/commit/3ee359f))



# [0.4.0](https://github.com/shimarulin/vuettext/compare/v0.3.0...v0.4.0) (2019-01-25)


### Features

* **msginit:** add msgmerge command ([1d7d758](https://github.com/shimarulin/vuettext/commit/1d7d758))
* **msgmerge:** update msgmerge command with globbing support ([2310625](https://github.com/shimarulin/vuettext/commit/2310625))



# [0.3.0](https://github.com/shimarulin/vuettext/compare/v0.2.1...v0.3.0) (2018-12-30)


### Bug Fixes

* **circleci:** install gettext ([7ef8b0b](https://github.com/shimarulin/vuettext/commit/7ef8b0b))
* **ts-node:** missing types ([1a2053a](https://github.com/shimarulin/vuettext/commit/1a2053a))
* **xgettext:** not informative output string ([a7665d9](https://github.com/shimarulin/vuettext/commit/a7665d9))


### Features

* **msginit:** add msginit command ([ee0a1d3](https://github.com/shimarulin/vuettext/commit/ee0a1d3))



## [0.2.1](https://github.com/shimarulin/vuettext/compare/v0.2.0...v0.2.1) (2018-12-25)



# [0.2.0](https://github.com/shimarulin/vuettext/compare/66df976...v0.2.0) (2018-12-25)


### Features

* **cli:** common parser, js parser ([22d807f](https://github.com/shimarulin/vuettext/commit/22d807f))
* **cli:** get file list by wildcard with fast-glob ([cc66867](https://github.com/shimarulin/vuettext/commit/cc66867))
* **cli:** replace yargs to oclif ([66df976](https://github.com/shimarulin/vuettext/commit/66df976))
* **cli:** xgettext command template ([27ee36c](https://github.com/shimarulin/vuettext/commit/27ee36c))
* **formatter:** add gettext formatter ([cba6422](https://github.com/shimarulin/vuettext/commit/cba6422))
* **parser:** add Vue parser ([a86ad52](https://github.com/shimarulin/vuettext/commit/a86ad52))
* **parser:** parse string literal ([f043437](https://github.com/shimarulin/vuettext/commit/f043437))


### Reverts

* **circleci:** test CircleCI branches ignore feature ([638e259](https://github.com/shimarulin/vuettext/commit/638e259))





src/commands/msginit.ts
import {Command, flags} from '@oclif/command'
import * as child_process from 'child_process'
import * as fuzzy from 'fuzzy'
import * as inquirer from 'inquirer'
import * as autocomplete from 'inquirer-autocomplete-prompt'
import * as iso6393 from 'iso-639-3'
import * as path from 'path'

import {mkpath} from '../common/mkpath'

inquirer.registerPrompt('autocomplete', autocomplete)

// tslint:disable-next-line
const searchLanguageCode = (answers: object, input = '') => {
  /**
   * ISO 639 implementation:
   * https://github.com/wooorm/iso-639-3
   *
   * More about language codes:
   * - https://www.gnu.org/software/gettext/manual/html_node/Language-Codes.html#Language-Codes
   * - https://www.gnu.org/software/gettext/manual/html_node/Header-Entry.html
   */
  const languages = iso6393
    .filter(el => el.type === 'living')
    .filter(lang => lang.scope !== 'special')
    .filter(lang => lang.iso6391 !== null)
    .sort((a, b) => {
      if (a.name > b.name) {
        return 1
      }
      if (a.name < b.name) {
        return -1
      }
      return 0
    })

  return new Promise(function (resolve) {
    let fuzzyResult = fuzzy.filter(input, languages, {
      extract(lang) {
        return input.length <= 3 && (lang.iso6391 === input || lang.iso6393 === input)
          ? lang.iso6391 || lang.iso6393 || ''
          : lang.name
      }
    })
    resolve(
      fuzzyResult.map(lang => {
        return lang.original.name
      })
    )
  })
}

const localeFlagDescription = `language code of the language. This can be in one of three forms:
- 'll', an ISO 639 two-letter language code (lowercase). See Language Codes for the list of codes.
- 'll_CC', where 'll' is an ISO 639 two-letter language code (lowercase) and 'CC' is an ISO 3166
  two-letter country code (uppercase). The country code specification is not redundant: Some languages
  have dialects in different countries. For example, 'de_AT' is used for Austria, and 'pt_BR' for Brazil.
  The country code serves to distinguish the dialects. See Language Codes and Country Codes for
  the lists of codes.
- 'll_CC@variant', where 'll' is an ISO 639 two-letter language code (lowercase), 'CC' is an ISO 3166
  two-letter country code (uppercase), and 'variant' is a variant designator. The variant designator
  (lowercase) can be a script designator, such as 'latin' or 'cyrillic'.
`

export default class Msginit extends Command {
  static description = 'Initialize language translation (*.po file for selected language)'

  static flags = {
    help: flags.help({
      char: 'h'
    }),
    locale: flags.string({
      char: 'l',
      description: localeFlagDescription,
    }),
    output: flags.string({
      char: 'o',
      description: 'Path to output file',
    }),
  }

  static args = [
    {
      name: 'file',
      required: true,
      description: 'Input *.pot file',
      default: 'i18n/messages.pot',
    }
  ]

  async run() {
    const {args, flags} = this.parse(Msginit)
    const {exec} = child_process

    const answers = await inquirer
      .prompt([
        {
          type: 'autocomplete',
          name: 'languageCode',
          // @ts-ignore
          suggestOnly: false,
          message: 'Select language',
          source: searchLanguageCode,
          filter: val => {
            const lang = iso6393.find(lang => lang.name === val)
            return lang && lang.iso6391
          },
          when: () => !flags.locale
        },
      ])

    const locale: string = flags.locale || answers.languageCode
    const output: string = flags.output || path.join(path.dirname(args.file), `locales/${locale}.po`)

    try {
      await mkpath(output)
      const outputMessage: string = await new Promise(resolve => {
        exec(
          `msginit --input=${args.file} --locale=${locale} --no-wrap --no-translator --output-file=${output}`,
          // tslint:disable-next-line:no-unused
          (err, stdout, stderr) => {
            if (err) throw err

            /**
             * Unexpected behavior: msginit always returns result to stderr
             */
            resolve(stderr.trim())
          })
      })
      this.log(outputMessage)
    } catch (e) {
      this.log(e)
    }
  }
}
